jobs/jobs.go

```diff
@@ -0,0 +1,235 @@
+package jobs
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/joyent/containerpilot/commands"
+	"github.com/joyent/containerpilot/discovery"
+	"github.com/joyent/containerpilot/events"
+)
+
+// Some magic numbers used internally by restart limits
+const (
+	unlimitedRestarts = -1
+	eventBufferSize   = 1000
+)
+
+// Job manages the state of a job and its start/stop conditions
+type Job struct {
+	Name             string
+	exec             *commands.Command
+	Status           bool // TODO: we'll need this to carry more info than bool
+	discoveryService discovery.Backend
+	Definition       *discovery.ServiceDefinition
+
+	// related events
+	startupEvent    events.Event
+	startupTimeout  time.Duration
+	stoppingEvent   events.Event
+	stoppingTimeout time.Duration
+
+	// timing and restarts
+	heartbeat      time.Duration
+	restartLimit   int
+	restartsRemain int
+	frequency      time.Duration
+
+	events.EventHandler // Event handling
+}
+
+// NewJob creates a new Job from a Config
+func NewJob(cfg *Config) *Job {
+	service := &Job{
+		Name:             cfg.Name,
+		exec:             cfg.exec,
+		heartbeat:        cfg.heartbeatInterval,
+		discoveryService: cfg.discoveryService,
+		Definition:       cfg.definition,
+		startupEvent:     cfg.startupEvent,
+		startupTimeout:   cfg.startupTimeout,
+		stoppingEvent:    cfg.stoppingEvent,
+		stoppingTimeout:  cfg.stoppingTimeout,
+		restartLimit:     cfg.restartLimit,
+		restartsRemain:   cfg.restartLimit,
+		frequency:        cfg.freqInterval,
+	}
+	service.Rx = make(chan events.Event, eventBufferSize)
+	service.Flush = make(chan bool)
+	if service.Name == "containerpilot" {
+		// TODO: right now this hardcodes the telemetry service to
+		// be always "on", but maybe we want to have it verify itself
+		// before heartbeating in the future
+		service.Status = true
+	}
+	return service
+}
+
+// FromConfigs creates Jobs from a slice of validated Configs
+func FromConfigs(cfgs []*Config) []*Job {
+	services := []*Job{}
+	for _, cfg := range cfgs {
+		service := NewJob(cfg)
+		services = append(services, service)
+	}
+	return services
+}
+
+// SendHeartbeat sends a heartbeat for this service
```

Comment:
Arguably not here either, as this is sends for service definition and not the job.